Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+    mutating func didYield(bufferDepth: Int) -> Bool
+
+    /// This method is called after the `Subscriber` consumed an element.
+    /// More specifically this method is called after `next` was called on an iterator of the ``NIOAsyncSequenceProducer``.
+    ///
+    /// - Parameter bufferDepth: The current depth of the internal buffer.
+    /// - Returns: Returns whether the producer should produce more elements.
+    mutating func didConsume(bufferDepth: Int) -> Bool
+}
+
+/// The delegate of ``NIOAsyncSequenceProducer``.
+public protocol NIOAsyncSequenceProducerDelegate: Sendable {
+    /// This method is called once the back-pressure strategy of the ``NIOAsyncSequenceProducer`` determined
+    /// that the producer needs to produce more elements.
+    ///
+    /// - Important: This is only called as a result of a `Subscriber` is calling ``NIOAsyncSequenceProducer/AsyncIterator/next()``.
+    /// It is never called as a result of a producer calling ``NIOAsyncSequenceProducer/Source/yield(_:)``.
+    func produceMore()
+
+    /// This method is called once the ``NIOAsyncSequenceProducer`` is terminated.
+    ///
+    /// Termination happens if:
+    /// - The ``NIOAsyncSequenceProducer/AsyncIterator`` is deinited.
+    /// - The ``NIOAsyncSequenceProducer`` deinited and no iterator is alive.
+    /// - The `Subscriber`'s `Task` is cancelled.
+    /// - The source finished and all remaining buffered elements have been consumed.
+    func didTerminate()
+}
+
+/// This is an ``AsyncSequence`` that supports a unicast ``AsyncIterator``.
+///
+/// The goal of this sequence is to produce a stream of elements from the _synchronous_ world
+/// (e.g. elements from a ``Channel`` pipeline) and vend it to the _asynchronous_ world for consumption.
+/// Furthermore, it provides facilities to implement a back-pressure strategy which
+/// observes the number of elements that are yielded and consumed. This allows to signal the source to request more data.
+///
+/// - Note: It is recommended to never directly expose this type from APIs, but rather wrap it. This is due to the fact that
+/// this type has three generic parameters where at least two should be known statically and it is really awkward to spell out this type.
+/// Moreover, having a wrapping type allows to optimize this to specialized calls if all generic types are known.
+///
+/// - Important: This sequence is a unicast sequence that only supports a single ``NIOAsyncSequenceProducer/AsyncIterator``.
+/// If you try to create more than one iterator it will result in a `fatalError`.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncSequenceProducer<
+    Element: Sendable,
+    Strategy: NIOAsyncSequenceProducerBackPressureStrategy,
+    Delegate: NIOAsyncSequenceProducerDelegate
+>: Sendable {
+    /// Simple struct for the return type of ``NIOAsyncSequenceProducer/makeSourceAndSequence(backPressureStrategy:delegate:)``.
+    public struct NewSequence {
```

Comment:
This should explain what the `sequence` and the `source` are for, add some paragraph about "use the `source` to yield values, and the `sequence` to consume them; the source can be freely shared with other parts of the program" or something like that

Reply:
Added some explanation on the properties of this type.